### ISSUE TITLE
about subgraph Gs

### ISSUE THREAD

**Issue by WuZongzhen** [Graph-COM/GSAT#4]
Thanks for your amazing work! But where does the code reflect the process of obtaining subgraphs (Gs) through sampling.

If you see this issue, please tell me the answer.Thanks in advance!

**Comment by siqim** [comment]
Hi Zongzhen, thanks for your questions!

In our implementation, we'll first obtain (unnormalized) node/edge attention weights by the line [here](https://github.com/Graph-COM/GSAT/blob/8865d078cfd5a3f3f79debf87d3da748238fe134/example/gsat.py#L38), where `sigmoid(att_log_logits)` essentially makes it the $p$ in our paper. To make the sampling process differentiable, we utilize the gumbel-softmax trick. So, `att` obtained [here](https://github.com/Graph-COM/GSAT/blob/8865d078cfd5a3f3f79debf87d3da748238fe134/example/gsat.py#L39) is the $\alpha \sim \operatorname{Bern}(p)$ in our paper.

But in practice, we don't yield $G_S$ by doing $\alpha \odot A$, because based on the gumbel-softmax trick it's non-trivial to make this operation differentiable for message-passing-based neural networks (MPNNs). Instead, no matter if `att` is node-level or edge-level attention, we first lift it to edge-level to obtain `edge_att`. After that, `edge_att` will act on the message of the corresponding edge, see [here](https://github.com/Graph-COM/GSAT/blob/8865d078cfd5a3f3f79debf87d3da748238fe134/src/models/conv_layers.py#L31). And this is like an approximation of $\alpha \odot A$ when using the gumbel-softmax trick for MPNNs.

Hope this could resolve your question. But feel free to let us know if you have other questions!

**Comment by WuZongzhen** [comment]
Thanks for your detailed reply, which helped me a lot.
